fairlearn/reductions/_exponentiated_gradient/_lagrangian.py

```diff
@@ -56,39 +56,39 @@ def __init__(self, X, sensitive_features, y, estimator, constraints, eps, B, opt
         self.last_linprog_n_hs = 0
         self.last_linprog_result = None
 
-    def _eval_from_error_gamma(self, error, gamma, lambda_vec):
+    def _eval(self, h, lambda_vec):
         """Return the value of the Lagrangian.
 
-        :return: tuple `(L, L_high)` where `L` is the value of the Lagrangian, and
-            `L_high` is the value of the Lagrangian under the best response of the lambda player
-        :rtype: tuple of two floats
+        :param h: TODO: should this be Q instead of h? It seems to indicate how much weight each h
+            gets in the randomized classifier, summing up to 1
+        :type h: pandas.Series
+        :param lambda_vec: lambda vector
+        :type h: pandas.Series
+
+        :return: tuple `(L, L_high, gamma, error)` where `L` is the value of the Lagrangian,
+            `L_high` is the value of the Lagrangian under the best response of the lambda player,
+            `gamma` is the vector of constraint violations, and `error` is the empirical error
         """
-        lambda_projected = self.constraints.project_lambda(lambda_vec)
+        error = self.errors[h.index].dot(h)
+        gamma = self.gammas[h.index].dot(h)
+
+        # TODO: some entries in gamma can be negative, presumably because for certain sensitive
```

Comment:
@MiroDudik  see the question in the code:
some entries in gamma can be negative, presumably because for certain sensitive feature values the (one-sided) constraint is satisfied and then some. However, below we multiply the lambda vector with it to calculate the Lagrangian value L, so negative values should reduce that. Is that intended? Just trying to make sure that's not unintended.

Reply:
Yes--it is totally fine for gamma to be negative. The formulas below are correct.